packages/offix-datastore/src/Model.ts

```diff
@@ -0,0 +1,72 @@
+import { Storage, StoreChangeEvent, EventTypes } from "./storage";
+import { createPredicate, Predicate } from "./predicates";
+
+export interface FieldOptions {
+    // GraphQL type
+    type: string;
+    // GraphQL key
+    key: string;
+    // TODO
+    format?: {  };
+}
+
+/**
+ * Defines the properties expected in the Fields object for a model
+ */
+export type Fields<T> = {
+    [P in keyof T]: FieldOptions
+};
+
+/**
+ * Provides CRUD capabilities for a model
+ */
+export class Model<T> {
+    private storeName: string;
+    private fields: Fields<T>;
+    private getStorage: () => Storage;
+
+    constructor(
+        storeName: string,
+        fields: Fields<T>,
+        getStorage: () => Storage
+    ) {
+        this.storeName = storeName;
+        this.fields = fields;
+        this.getStorage = getStorage;
+    }
+
+    public save(input: T): Promise<T> {
+        return this.getStorage().save(this.storeName, input);
+    }
+
+    public query(predicateFunction?: Predicate<T>) {
+        if (!predicateFunction) {return this.getStorage().query(this.storeName);}
+
+        const modelPredicate = createPredicate(this.fields);
+        const predicate = predicateFunction(modelPredicate);
+        return this.getStorage().query(this.storeName, predicate);
+    }
+
+    public update(input: Partial<T>, predicateFunction?: Predicate<T>) {
+        if (!predicateFunction) {
+            return this.getStorage().update(this.storeName, input);
+        }
+
+        const modelPredicate = createPredicate(this.fields);
+        const predicate = predicateFunction(modelPredicate);
+        return this.getStorage().update(this.storeName, input, predicate);
+    }
+
+    public remove(predicateFunction?: Predicate<T>) {
+        if (!predicateFunction) {return this.getStorage().remove(this.storeName);}
+
+        const modelPredicate = createPredicate(this.fields);
+        const predicate = predicateFunction(modelPredicate);
```

Comment:
Why do we have predicate for deletion?

Reply:
Would users will need to define their id?
Is possible to delete by username or something like that?

Reply:
Predicate allows them to select by anything.
So they can delete by id, by title, by title and description etc.